core/controllers/editor.py

```diff
@@ -739,12 +747,17 @@ def put(self, exploration_id):
         exp_user_data = exp_services.get_user_exploration_data(
             self.user_id, exploration_id)
         # If the draft_change_list_id is False, have the user discard the draft
-        # changes. We save the draft to the datastore even if the version is
-        # invalid, so that it is available for recovery later.
+        # changes. We save the draft to the datastore only when the changes are
+        # mergeable even if the version is invalid and if they are not mergeable
+        # we discard them and show them in the frontend discard changes modal
+        # and give an option to the user to export them so that it is available
+        # for recovery later.
         self.render_json({
             'draft_change_list_id': exp_user_data['draft_change_list_id'],
             'is_version_of_draft_valid': exp_services.is_version_of_draft_valid(
```

Comment:
Yes, because changes_are_mergeable returns true when the versions are the same. So we just need to check sometimes that changes_are_mergeable returns true when the versions are invalid like we did on line 117 of change-list.service.ts